README
NAME
    App::CPANIDX::HTTP::Server - HTTP::Server::Simple based server for
    CPANIDX

VERSION
    version 0.06

SYNOPSIS
      use strict;
      use warnings;
      use App::CPANIDX::HTTP::Server;

      my $dsn = 'dbi:SQLite:dbname=cpanidx.db';
      my $user = '';
      my $pass = '';
      my $port = 8082; # the port to listen for requests on

      my $server = App::CPANIDX::HTTP::Server->new( $port );
      $server->dsn( $dsn, $user, $pass );
      $server->run();

      # Requests can now be directed to http://nameofyourserver:8082/cpanidx/

DESCRIPTION
    App::CPANIDX::HTTP::Server is a HTTP::Server::Simple based server for
    CPANIDX. Use the "cpanidx-gendb" script provided by App::CPANIDX to
    generate a CPANIDX database and then use this module to serve the
    associated data.

METHODS
    "new"
        Start a new instance of App::CPANIDX::HTTP::Server. Takes one
        option, the port number to start listening on for requests. If it is
        not provided will default to 8080.

    "dsn"
        After running "new", but before calling "run", call this to assign
        the database details to the server. Takes three arguments: a DBI
        "DSN" string, a username (if applicable) and a password (if
        applicable).

    "run"
        Runs the server and starts handling requests.

    "handle_request"
        Deals with requests. No user serviceable parts.

SEE ALSO
    App::CPANIDX

AUTHOR
    Chris Williams <chris@bingosnet.co.uk>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2011 by Chris Williams.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
==============================
2011-01-03 14:31:33 +0000 0.06
==============================

  commit f1e666eab93b8228894bc083383d989731944357
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Mon Jan 3 14:31:33 2011 +0000
  
    School-boy errors fixed.

  commit 83bbf316018ed24bcfab73db454879176bdba351
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Mon Jan 3 14:27:11 2011 +0000
  
    Make the error handling more robust so it doesn't kill the server 

==============================
2011-01-01 22:33:53 +0000 0.04
==============================

  commit 71ece539b5b1721a05a5fd57bd1ec697e551dfdb
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Sat Jan 1 22:33:53 2011 +0000
  
    Should be requiring 0.28

  commit ccfb621f3f5ce4baebed5d0eb07e9d834e270916
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Sat Jan 1 22:32:35 2011 +0000
  
    Updates for new App::CPANIDX 

====================================
End of changes in the last 1825 days
====================================


lib/App/CPANIDX/HTTP/Server.pm
package App::CPANIDX::HTTP::Server;
BEGIN {
  $App::CPANIDX::HTTP::Server::VERSION = '0.06';
}
 
#ABSTRACT: HTTP::Server::Simple based server for CPANIDX

use strict;
use warnings;
use DBI;
use App::CPANIDX::Renderer;
use App::CPANIDX::Queries;
use HTTP::Server::Simple::CGI;
use base qw(HTTP::Server::Simple::CGI);
 
sub dsn {
  my ($self,$dsn,$user,$pass) = @_;
  if ( $dsn and $self->{_dbh} ) {
    warn "Already have a database connection, thanks\n";
    return;
  }
  if ( $dsn ) {
    $self->{_dbh} = DBI->connect($dsn,$user,$pass) or die $DBI::errstr, "\n";
    $self->{_dsn} = $dsn;
    return;
  }
  return $self->{_dsn};
}

sub handle_request {
   my $self = shift;
   my $cgi  = shift;
   
   my $path = $cgi->path_info();
   $path =~ s!/+!/!g;
   warn $path, "\n";
   my ($root,$enc,$type,$search) = grep { $_ } split m#/#, $path;

   if ( $root eq 'cpanidx' and $enc and $type ) {
      $search = '0' if $type =~ /^next/ and !$search;
      my @results = $self->_search_db( $type, $search );
      $enc = 'yaml' unless $enc and grep { lc($enc) eq $_ } App::CPANIDX::Renderer->renderers();
      my $ren = App::CPANIDX::Renderer->new( \@results, $enc );
      my ($ctype, $string) = $ren->render( $type );
      print "HTTP/1.0 200 OK\r\n";
      print "Content-type: $ctype\r\n\r\n";
      print $string;
   } 
   else {
      print "HTTP/1.0 404 Not found\r\n";
      print $cgi->header,
      $cgi->start_html('Not found'),
      $cgi->h1('Not found'),
      $cgi->end_html;
   }
}

sub _search_db {
  my ($self,$type,$search) = @_;
  my @results;
  if ( my $sql = App::CPANIDX::Queries->query( $type ) ) {
    if ( ( $type eq 'mod' or $type eq 'corelist' or $type eq 'perms' ) 
        and !( $search =~ m#\A[a-zA-Z_][0-9a-zA-Z_]*(?:(::|')[0-9a-zA-Z_]+)*\z# ) ) {
      return @results;
    } 
    # send query to dbi
    if ( my $sth = $self->{_dbh}->prepare_cached( $sql->[0] ) ) {
      $sth->execute( ( $sql->[1] ? $search : () ) );
      while ( my $row = $sth->fetchrow_hashref() ) {
        push @results, { %{ $row } };
      }
    }
    else {
      warn $DBI::errstr, "\n";
      return @results;
    }
  }
  return @results;
}

1;


__END__
=pod

=head1 NAME

App::CPANIDX::HTTP::Server - HTTP::Server::Simple based server for CPANIDX

=head1 VERSION

version 0.06

=head1 SYNOPSIS

  use strict;
  use warnings;
  use App::CPANIDX::HTTP::Server;

  my $dsn = 'dbi:SQLite:dbname=cpanidx.db';
  my $user = '';
  my $pass = '';
  my $port = 8082; # the port to listen for requests on

  my $server = App::CPANIDX::HTTP::Server->new( $port );
  $server->dsn( $dsn, $user, $pass );
  $server->run();

  # Requests can now be directed to http://nameofyourserver:8082/cpanidx/

=head1 DESCRIPTION

App::CPANIDX::HTTP::Server is a L<HTTP::Server::Simple> based server for CPANIDX.
Use the C<cpanidx-gendb> script provided by L<App::CPANIDX> to generate a CPANIDX
database and then use this module to serve the associated data.

=head1 METHODS

=over

=item C<new>

Start a new instance of App::CPANIDX::HTTP::Server. Takes one option, the port number to
start listening on for requests. If it is not provided will default to C<8080>.

=item C<dsn>

After running C<new>, but before calling C<run>, call this to assign the database details to 
the server. Takes three arguments: a L<DBI> C<DSN> string, a username (if applicable) and a 
password (if applicable).

=item C<run>

Runs the server and starts handling requests.

=item C<handle_request>

Deals with requests. No user serviceable parts.

=back

=head1 SEE ALSO

L<App::CPANIDX>

=head1 AUTHOR

Chris Williams <chris@bingosnet.co.uk>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Chris Williams.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;

BEGIN { require 5.006; }

use ExtUtils::MakeMaker 6.31;



my %WriteMakefileArgs = (
  'ABSTRACT' => 'HTTP::Server::Simple based server for CPANIDX',
  'AUTHOR' => 'Chris Williams <chris@bingosnet.co.uk>',
  'BUILD_REQUIRES' => {
    'Test::More' => '0.47'
  },
  'CONFIGURE_REQUIRES' => {
    'ExtUtils::MakeMaker' => '6.31'
  },
  'DISTNAME' => 'App-CPANIDX-HTTP-Server',
  'EXE_FILES' => [
    'bin/cpanidx_httpserver'
  ],
  'LICENSE' => 'perl',
  'NAME' => 'App::CPANIDX::HTTP::Server',
  'PREREQ_PM' => {
    'App::CPANIDX' => '0.28',
    'Config::Tiny' => '2.12',
    'HTTP::Server::Simple' => '0.43',
    'HTTP::Server::Simple::CGI' => '0'
  },
  'VERSION' => '0.06',
  'test' => {
    'TESTS' => 't/*.t'
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



